### DISCUSSION TITLE
Updating NPC (hp, name, etc)

### DISCUSSION THREAD

**Discussion by Maklock** [Pryaxis/TShock#2108] [Q&A]
I'm still learning how plugins work, this is how my command looks currently:

```
private void Test(CommandArgs args)
        {
            int spawnTileX;
            int spawnTileY;
            TShock.Utils.GetRandomClearTileWithInRange(args.Player.TileX, args.Player.TileY, 50, 20, out spawnTileX, out spawnTileY);

            int index = NPC.NewNPC(spawnTileX * 16, spawnTileY * 16, 302);
            NPC boss = Main.npc[index];

            boss.GivenName = "Peko";
            boss._givenName = "Peko";
            boss.lifeMax = 10000;
            boss.life = 10000;
            boss.aiStyle = 60;
            
            NetMessage.SendData((int)PacketTypes.NpcUpdate, -1, -1, NetworkText.Empty, index);
            NetMessage.SendData((int)PacketTypes.UpdateNPCName, -1, -1, NetworkText.Empty, index);
        }
```

So I've been just testing out what I can change,
with the code above a Halloween Slime will spawn with the name Peko,
it does have 10.000 HP but, it will only show once you hit the monster once,
and it will be displayed as 10000/127, so those values are not updating on the client.
AI is wonky, I don't know if that is the right way to change an NPC's AI,
it does change, but it's weird.. like a mix between both the normal slime AI and the new one.

Am I sending NpcUpdate wrong?

**Comment by bartico6** [comment]
The packet doesn't always include the hp value, and it never includes maxhealth. Here's how it works:

The game assumes the maxhealth is the same on both client and server (because it's not supposed to be modified) so when you're at 10k/10k, the game simply says "this npc is at max hp" and that's all the client is told, resulting in 127/127hp. Once the NPC takes damage, it's no longer at full health, so the real health value is sent, resulting in 10k/127.

Don't touch the maxhealth value and it'll work correctly.

As for name: you're doing it right.

**Reply by Maklock** [reply]
Ohhhh!
That actually makes a lot of sense, so basically it just sets lifeMax to the starting life value.

**Reply by bartico6** [reply]
Not quite.
What it does is it makes client set **clientside current hp** to **clientside max hp** when **serverside current hp** equals **serverside max hp**. If you change the serverside lifeMax, you'll have issues where the client will perceive incorrect health values. Just don't touch the lifeMax variable and instead use current health only.
